server/backends/migration_cache/migration_cache.go

```diff
@@ -121,46 +132,160 @@ func pebbleCacheFromConfig(env environment.Env, cfg *PebbleCacheConfig) (*pebble
+func (mc *MigrationCache) WithIsolation(ctx context.Context, cacheType interfaces.CacheType, remoteInstanceName string) (interfaces.Cache, error) {
 	return nil, status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) Contains(ctx context.Context, d *repb.Digest) (bool, error) {
+func (mc *MigrationCache) Contains(ctx context.Context, d *repb.Digest) (bool, error) {
 	return false, status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) Metadata(ctx context.Context, d *repb.Digest) (*interfaces.CacheMetadata, error) {
+func (mc *MigrationCache) Metadata(ctx context.Context, d *repb.Digest) (*interfaces.CacheMetadata, error) {
 	return nil, status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) FindMissing(ctx context.Context, digests []*repb.Digest) ([]*repb.Digest, error) {
+func (mc *MigrationCache) FindMissing(ctx context.Context, digests []*repb.Digest) ([]*repb.Digest, error) {
 	return nil, status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) Get(ctx context.Context, d *repb.Digest) ([]byte, error) {
+func (mc *MigrationCache) GetMulti(ctx context.Context, digests []*repb.Digest) (map[*repb.Digest][]byte, error) {
 	return nil, status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) GetMulti(ctx context.Context, digests []*repb.Digest) (map[*repb.Digest][]byte, error) {
-	return nil, status.UnimplementedError("not yet implemented")
-}
-
-func (mc MigrationCache) Set(ctx context.Context, d *repb.Digest, data []byte) error {
+func (mc *MigrationCache) SetMulti(ctx context.Context, kvs map[*repb.Digest][]byte) error {
 	return status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) SetMulti(ctx context.Context, kvs map[*repb.Digest][]byte) error {
+func (mc *MigrationCache) Delete(ctx context.Context, d *repb.Digest) error {
 	return status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) Delete(ctx context.Context, d *repb.Digest) error {
-	return status.UnimplementedError("not yet implemented")
+func (mc *MigrationCache) Reader(ctx context.Context, d *repb.Digest, offset, limit int64) (io.ReadCloser, error) {
+	return nil, status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) Reader(ctx context.Context, d *repb.Digest, offset, limit int64) (io.ReadCloser, error) {
+func (mc *MigrationCache) Writer(ctx context.Context, d *repb.Digest) (io.WriteCloser, error) {
 	return nil, status.UnimplementedError("not yet implemented")
 }
 
-func (mc MigrationCache) Writer(ctx context.Context, d *repb.Digest) (io.WriteCloser, error) {
-	return nil, status.UnimplementedError("not yet implemented")
+type getResult struct {
+	fromSrcCache bool
+	data         []byte
+	err          error
+}
+
+// TODO(Maggie): Add copying logic
+func (mc *MigrationCache) Get(ctx context.Context, d *repb.Digest) ([]byte, error) {
```

Comment:
another way to structure this might be using an errgroup. I think it's a little clearer than the channel option you propose here, but maybe I'm overlooking something critical about it. Here it is so you can consider it:

```
eg, gctx := errgroup.WithContext(ctx)
defer gctx.Cancel()

doubleRead := rand.Float64() < doubleReadPercentage
var srcErr, dstErr error
var srcBuf []byte

eg.Go(func() error {
  srcBuf, srcErr = mc.src.Get(gctx, d)
  return srcErr
})

if doubleRead {
  eg.Go(func() error {
    _, dstErr = mc.dst.Get(gctx, d)
    return nil // we don't care about the return error from this cache
  })
}

if err := eg.Wait(); err != nil {
  return nil, err
}

if doubleRead {
  if srcErr != dstErr {
    log.Warningf("Double read of %q failed. Src returned %s, Dest returned %s", d, srcErr, dstErr)
  }
}

return srcBuf, srcErr
```

Reply:
That's clearer thanks. Updated to this